internal/app/fipcontroller/controller.go

```diff
@@ -81,34 +81,37 @@ func (controller *Controller) Run(ctx context.Context) error {
 	}
 }
 
-/*
- * Main logical function.
- *  Searches for the Hetzner Cloud Node object the pod is running on and validates that all configured floating IPs
- *  are attached to that node.
- */
+// UpdateFloatingIPs searches for running hetzner cloud servers and sort them by fewest assigned floating ips.
+// It then (re)assigns all unassigned ips or ips that are assigned to non running servers to the sorted running serves.
 func (controller *Controller) UpdateFloatingIPs(ctx context.Context) error {
 	controller.Logger.Debugf("Checking floating IPs")
 
-	nodeAddress, err := controller.nodeAddress(ctx, controller.Configuration.NodeName, controller.Configuration.NodeAddressType)
+	// Get running servers for floating ip assignment
+	nodeAddressList, err := controller.nodeAddressList(ctx, controller.Configuration.NodeAddressType)
 	if err != nil {
-		return fmt.Errorf("could not get kubernetes node address: %v", err)
+		return fmt.Errorf("could not get addressList for active kubernetes nodes: %v", err)
 	}
-	controller.Logger.Debugf("Found node address: %s", nodeAddress.String())
 
-	server, err := controller.server(ctx, nodeAddress)
+	runningServers, err := controller.servers(ctx, nodeAddressList)
 	if err != nil {
-		return fmt.Errorf("could not get configured server: %v", err)
+		return fmt.Errorf("Could not get server objects for addressList: %v", err)
 	}
-	controller.Logger.Debugf("Found server: %s (%d)", server.Name, server.ID)
 
-	for _, floatingIPAddr := range controller.Configuration.HcloudFloatingIPs {
-		floatingIP, err := controller.floatingIP(ctx, floatingIPAddr)
-		if err != nil {
-			return fmt.Errorf("could not get floating IP '%s': %v", floatingIPAddr, err)
-		}
+	// Get floatingIPs from config if specified, otherwise from hetzner api
+	floatingIPs, err := controller.getFloatingIPs(ctx)
+	if err != nil {
+		return fmt.Errorf("Could not get floatingIPs: %v", err)
+	}
+
+	for _, floatingIP := range floatingIPs {
 		controller.Logger.Debugf("Checking floating IP: %s", floatingIP.IP.String())
 
-		if floatingIP.Server == nil || server.ID != floatingIP.Server.ID {
+		// (Re)assign floatingIP if no server is assigned or the assigned server is not running
+		// Since we already have all running server in a slice we can just search through it
+		if floatingIP.Server == nil || !findServerByID(runningServers, floatingIP.Server) {
+			// Get the server with the lowest amount of fips (cant be nil since we know that servers can't be empty)
```

Comment:
Are you sure about your assumption? Since `floatingIP.Server == nil` is a valid case `findServerByID()` might not be executed. For `controller.servers` I also don't see a 100% safeguard; e.g. assume multiple hcloud accounts and messed up tokens, IP and servers from different accounts.. This might still not be working, but please add the nil-check here (it doesn't hurt and prevents a failure in case of bugs [be it ours, hetzners, or someone else's])

Reply:
I am not sure what you mean exactly. If `floatingIP.Server == nil` then `findServerByID()` should not be executed, because we already know, that we need to assign this ip to a hetzner vm.

And if the first condition is not correct we know for sure, that `floatingIP.Server != nil`

Reply:
Reading over this again I'm not sure either what I meant with the first part. Maybe I just misread something.
The second part was a comment on the `cant be nil since we know that servers can't be empty` assumption. Please validate that anyhow, since it doesn't hurt but might catch errors/bugs without crashing the controller

Reply:
okay now I got you.
I have added nil and emtpy list checks for both the kuberentes ip slice and the hetzner server vm slice.